File: parser/datacap-parser-trino/src/main/java/io/edurt/datacap/parser/node/NodeLocation.java
package io.edurt.datacap.parser.node;

import java.util.Objects;

import static com.google.common.base.Preconditions.checkArgument;

public final class NodeLocation
{
    private final int line;
    private final int charPositionInLine;

    public NodeLocation(int line, int charPositionInLine)
    {
        checkArgument(line >= 1, "line must be at least one, got: %s", line);
        checkArgument(charPositionInLine >= 0, "charPositionInLine must be at least zero, got: %s", charPositionInLine);
        this.line = line;
        this.charPositionInLine = charPositionInLine;
    }

    public int getLineNumber()
    {
        return line;
    }

    public int getColumnNumber()
    {
        return charPositionInLine + 1;
    }

    @Override
    public String toString()
    {
        return line + ":" + charPositionInLine;
    }

    @Override
    public boolean equals(Object o)
    {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        NodeLocation that = (NodeLocation) o;
        return line == that.line &&
                charPositionInLine == that.charPositionInLine;
    }

    @Override
    public int hashCode()
    {
        return Objects.hash(line, charPositionInLine);
    }
}

File: parser/datacap-parser-mysql/src/main/kotlin/io/edurt/datacap/parser/mysql/companion/ast/NodeLocation.kt
package io.edurt.datacap.parser.mysql.companion.ast

import com.google.common.base.Preconditions.checkArgument
import java.util.*

data class NodeLocation(val line: Int, val charPositionInLine: Int) {
    init {
        checkArgument(line >= 1, "line must be at least one, got: $line")
        checkArgument(charPositionInLine >= 0, "charPositionInLine must be at least zero, got: $charPositionInLine")
    }

    fun getLineNumber(): Int {
        return line
    }

    fun getColumnNumber(): Int {
        // Since the index is 0, perform +1 operation here
        return charPositionInLine + 1
    }

    override fun toString(): String {
        return "$line:$charPositionInLine"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as NodeLocation
        return line == that.line && charPositionInLine == that.charPositionInLine
    }

    override fun hashCode(): Int {
        return Objects.hash(line, charPositionInLine)
    }
}
